plat/kvm/arm/smp.c

```diff
@@ -0,0 +1,401 @@
+		return -1;
+	}
+
+	/* Get size cells */
+	size_prop = fdt_getprop(_libkvmplat_cfg.dtb, fdt_cpu,
+			"#size-cells", &cell);
+	if (size_prop == NULL) {
+		naddr = FDT_SIZE_CELLS_DEFAULT;
+		uk_pr_warn("Using default value of 0 for cpu size-cells.\n");
+	} else {
+		size = fdt32_to_cpu(size_prop[0]);
+	}
+	if (size != 0) {
+		uk_pr_err("Invalid size-cells size for cpu node!\n");
+		return -1;
+	}
+
+	/* Search all the cpu nodes in DTB */
+	cpu_idx = 0;
+	fdt_for_each_subnode(subnode, _libkvmplat_cfg.dtb, fdt_cpu) {
+		int prop_len = 0;
+		__u32 cpu_id;
+
+		cpu_prop = fdt_get_property(_libkvmplat_cfg.dtb, subnode,
+						"device_type", &prop_len);
+		if (!cpu_prop)
+			continue;
+		if (strcmp(cpu_prop->data, "cpu"))
+			continue;
+
+		cpu_prop = fdt_get_property(_libkvmplat_cfg.dtb, subnode,
+						"enable-method", NULL);
+		if (!cpu_prop) {
+			uk_pr_err("enable-method field is missing \
+				in cpu fdt node!\n");
+			continue;
+		} else if (strcmp(cpu_prop->data, "psci")) {
+			uk_pr_err("enable-method is not PSCI in \
+				cpu fdt node: (%s)\n",
+				cpu_prop->data);
+			continue;
+		}
+
+		cpu_prop = fdt_get_property(_libkvmplat_cfg.dtb, subnode,
+						"reg", &prop_len);
+		if (cpu_prop == NULL || prop_len <= 0) {
+			uk_pr_err("Error when reading cpu reg field \
+				(cpu id) in fdt!\n");
+			continue;
+		}
+
+		id_reg = (const fdt32_t *) cpu_prop->data;
+		cpu_id = fdt_reg_read_number(id_reg, naddr);
+		if (cpu_id != bsp_cpu_id) {
+			cpu_idx++;
+			cpus[cpu_idx].id = cpu_id;
+			ukarch_spin_init(&cpus[cpu_idx].lock);
+			uk_pr_info("Initialized core (%u)\n",
+				cpus[cpu_idx].id);
+		}
+	}
+	smp_numcores = cpu_idx + 1;
+
+	return 0;
+}
+
+static int psci_cpu_on(struct arm64_cpu *cpu, __uptr entry)
+{
+	/* PSCI v0.1 and v0.2 both support cpu_on. the parameter @cpu after
+	 * @entry will be passed as x0 for entry
+	 */
+	return smcc_psci_hvc_call(PSCI_FNID_CPU_ON, (__u64) cpu->id,
+					(__u64) entry, (__u64) cpu);
+}
+
+int ukplat_lcpu_start(__lcpuid lcpuid[], void *sp[],
+				ukplat_lcpu_entry_t entry[], unsigned int num)
+{
+	__uptr lcpu_start_addr = ukplat_virt_to_phys(_lcpu_start);
+	__u32 this_cpu_id = ukplat_lcpu_id();
+	__lcpuid cpu_idx;
+	__u32 i, n = 0;
+	int err;
+
+	UK_ASSERT(sp);
+
+	if (num == 0) {
+		UK_ASSERT(!lcpuid);
+		num = smp_numcores;
+	}
+
+
+	/* register the lcpu_run and lcpu_wakeup interrupt handlers */
+	err = ukplat_irq_register(UKPLAT_ARM64_IPI_RUN_IRQ,
+					arm64_lcpu_ipi_run_handler, NULL);
+	if (err != 0) {
+		uk_pr_crit("Interrupt handler for IRQ %d failed\n",
+			UKPLAT_ARM64_IPI_RUN_IRQ);
+		return err;
+	}
+
+	/* start the secondary cores */
+	for (i = 0; i < num; i++) {
+		cpu_idx = (lcpuid) ? lcpuid[i] : (__lcpuid) i;
+
+		if (cpus[cpu_idx].id == this_cpu_id)
+			continue;
+
+		if (cpu_idx >= smp_numcores)
+			continue;
+
+		if (cpus[cpu_idx].state != LCPU_STATE_OFFLINE)
+			uk_pr_warn("Core %d already starting, \
+				moving on.\n", cpu_idx);
+
+		if (ukarch_spin_trylock(&cpus[cpu_idx].lock)) {
```

Comment:
For now, yes. However, we've discussed about having a more powerful boot process, where, for instance, one core starts two cores and then the new started cores can start other ones. Also, right now nothing prevents the user from calling `ukplat_lcpu_start()` multiple times, from multiple cores. Therefore, conflicts can appear, with 2 cores trying to start the same third core.
Using `trylock()`, we check if the lock associated to one core is already acquired, meaning that it is already in the process of being started. Therefore, if trylock failed, we know we can skip the targeted core, since it is already being started by another one.

Reply:
Ah okay. I wasn't aware of that discussion. Thanks 👍🏼